test/unit/org/apache/cassandra/db/tries/IntersectionTrieTest.java

```diff
@@ -0,0 +1,518 @@
+                             TrieSet.range(VERSION, before(4), before(7)));
+            // touching, i.e. still non-overlapping
+            testIntersection("", asList(), trie,
+                             TrieSet.range(VERSION, before(0), before(3)),
+                             TrieSet.range(VERSION, before(3), before(7)));
+            // overlapping 1
+            testIntersection("", asList(2), trie,
+                             TrieSet.range(VERSION, before(0), before(3)),
+                             TrieSet.range(VERSION, before(2), before(7)));
+            // overlapping 2
+            testIntersection("", asList(1, 2), trie,
+                             TrieSet.range(VERSION, before(0), before(3)),
+                             TrieSet.range(VERSION, before(1), before(7)));
+            // covered
+            testIntersection("", asList(0, 1, 2), trie,
+                             TrieSet.range(VERSION, before(0), before(3)),
+                             TrieSet.range(VERSION, before(0), before(7)));
+            // covered 2
+            testIntersection("", asList(1, 2), trie,
+                             TrieSet.range(VERSION, before(1), before(3)),
+                             TrieSet.range(VERSION, before(0), before(7)));
+        }
+    }
+
+    @Test
+    public void testSimpleRanges() throws TrieSpaceExhaustedException
+    {
+        for (bits = bitsNeeded; bits > 0; --bits)
+        {
+            Trie<Integer> trie = fromList(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+
+            testIntersection("", asList(0, 1, 2, 3, 4, 5, 6, 7, 8, 9), trie);
+
+            testIntersection("", asList(3, 4, 5, 6), trie,
+                             TrieSet.ranges(VERSION, before(3), before(7)));
+
+            testIntersection("", asList(3), trie,
+                             TrieSet.ranges(VERSION, before(3), before(4)));
+
+            testIntersection("", asList(0, 1, 2, 3, 4, 5, 6), trie,
+                             TrieSet.ranges(VERSION, null, before(7)));
+
+            testIntersection("", asList(3, 4, 5, 6, 7, 8, 9), trie,
+                             TrieSet.ranges(VERSION, before(3), null));
+
+            testIntersection("", asList(0, 1, 2, 3, 4, 5, 6, 7, 8, 9), trie,
+                             TrieSet.ranges(VERSION, null, null));
+
+            testIntersection("", asList(3, 4, 5, 7, 8), trie,
+                             TrieSet.ranges(VERSION, before(3), before(6), before(7), before(9)));
+
+            testIntersection("", asList(3, 7, 8), trie,
+                             TrieSet.ranges(VERSION, before(3), before(4), before(7), before(9)));
+
+            testIntersection("", asList(3, 7, 8), trie,
+                             TrieSet.ranges(VERSION, before(3), before(4), before(7), before(9), before(12), before(15)));
+
+            testIntersection("", asList(3, 4, 5, 6, 7, 8), trie,
+                             TrieSet.ranges(VERSION, before(3), before(9)));
```

Comment:
Same as the test for [before(3),before(7)] ?

Reply:
Removed